# NGS-Variant-Discovery-Pipeline-GATK
## Note - Workflow inprogress

This repository demonstrates a real-world, end-to-end NGS variant analysis pipeline, starting from raw sequencing reads to align, preprocess, discover variants, filter quality, annotate functionality, and validate results.

The pipeline includes:

High-quality read alignment to a reference genome

Sorting, duplicate marking, and base quality recalibration

Variant calling using industry-standard tools

Variant filtering and quality assessment

Functional annotation to interpret biological impact

Visualization for manual verification of variant regions

This specific workflow and related material provided by the SIB Swiss Institute of Bioinformatics, and all steps are executed in a reproducible manner using scripted workflows and controlled software environments.
